### PR TITLE
Matrix Testing

### PR DESCRIPTION
**Pull request by hliuson**
# Description

This pull request implements matrix testing, which extends integration testing in a handful of ways:
- Given a Pulumi YAML or PCL program, matrix testing automatically converts it into Pulumi programs in different languages and perform integration testing on them.
- Matrix testing can make use of the plugin link feature and ensure that the tested projects use the correct plugin binary. It additionally performs codegen using the plugin's schema and tests the generated projects using the generated SDKs from the schema.

Codegen is modified minorly to make this happen, as well as to enable specific testing of the core pulumi/pulumi language SDKs.

It additionally adds auxiliary stacks to integration tests which may be used to test a pulumi stack which depends on another stack via a stack reference.

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [x] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/codegen/go/gen_program.go [hunk 1]
```diff
@@ -145,21 +146,34 @@ func GenerateProject(directory string, project workspace.Project, program *pcl.P
...
+				}
+			}
+		}
+	}
+
 	gomod.WriteString(`
-go 1.17
 
 require (
 	github.com/pulumi/pulumi/sdk/v3 v3.30.0
```
Comment (Frassle) [inline pkg/codegen/go/gen_program.go:173]: Oh god's this is hard coded to 3.30! Pretty sure this should be floating latest like for dotnet.

File: pkg/codegen/go/gen_program.go [hunk 2]
```diff
@@ -145,21 +146,34 @@ func GenerateProject(directory string, project workspace.Project, program *pcl.P
...
 	var gomod bytes.Buffer
 	gomod.WriteString("module " + project.Name.String() + "\n")
+	gomod.WriteString("go 1.17\n")
+	for name, sdk := range localProjects {
+		if name == "go" {
+			gomod.WriteString(fmt.Sprintf("replace github.com/pulumi/pulumi/sdk/v3/go => %s\n", sdk))
+		} else {
+			for _, p := range packages {
+				if p.Name == name {
+					gomod.WriteString(fmt.Sprintf("replace github.com/pulumi/pulumi-%s/sdk/go => %s/go\n", name, sdk))
```
Comment (Frassle) [inline pkg/codegen/go/gen_program.go:164]: Similar to dotnet, I think it should be on the caller of this function to provide the full package paths

File: pkg/codegen/nodejs/gen.go [hunk 3]
```diff
@@ -2183,6 +2182,10 @@ func genNPMPackageMetadata(pkg *schema.Package, info NodePackageInfo) string {
 		},
 	}
 
+	if !pkg.Test {
```
Comment (Frassle) [inline pkg/codegen/nodejs/gen.go:2185]: Add a comment about why this isn't included when Test is set

File: pkg/codegen/python/gen.go [hunk 4]
```diff
@@ -1981,34 +1981,43 @@ func genPackageMetadata(
 	pluginVersion := version
 	info, ok := pkg.Language["python"].(PackageInfo)
 	if pkg.Version != nil && ok && info.RespectSchemaVersion {
-		version = pypiVersion(*pkg.Version)
+		version = pypiVersion(*pkg.Version) //Does not generate correct version names for pre-release versions.
```
Comment (Frassle) [inline pkg/codegen/python/gen.go:1984]: Oh? Can we fix that? What's the problem?

File: pkg/testing/integration/program.go [hunk 5]
```diff
@@ -1060,9 +1073,36 @@ func (pt *ProgramTester) TestCleanUp() {
 	}
 }
 
+//TODO: Allow matrix tester to pre-initialize stacks to avoid redundant pulumi ups.
```
Comment (Frassle) [inline pkg/testing/integration/program.go:1076]: Still TODO here?

File: pkg/testing/matrix/matrix.go [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+	}
+	assert.NotNil(t, proj)
+	assert.NotNil(t, pclProgram)
+
+	localProjects := map[string]string{}
+
+	//Here we are overriding the plugin links to use our own plugin links.
+	proj.Plugins = &workspace.Plugins{}
+
+	//This needs further testing
```
Comment (Frassle) [inline pkg/testing/matrix/matrix.go:140]: ?

File: pkg/testing/matrix/matrix.go [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+		loader := schema.NewPluginLoader(ctx.Host)
+		_, pclProgram, err = yamlgen.Eject(pwd, loader)
+		assert.NoError(t, err)
+	}
+	assert.NotNil(t, proj)
+	assert.NotNil(t, pclProgram)
+
+	localProjects := map[string]string{}
+
+	//Here we are overriding the plugin links to use our own plugin links.
```
Comment (Frassle) [inline pkg/testing/matrix/matrix.go:137]: Assert it's empty before setting it?

File: pkg/testing/matrix/matrix.go [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+	proj.Plugins = &workspace.Plugins{}
+
+	//This needs further testing
+	for lang, path := range opts.PulumiSDKs {
+		localProjects[lang] = path
```
Comment (Frassle) [inline pkg/testing/matrix/matrix.go:142]:
```suggestion
		localProjects["pulumi"] = path
```

File: pkg/testing/matrix/matrix.go [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+		var spec schema.PackageSpec
+		err = json.Unmarshal(schemaBytes, &spec)
+		assert.NoError(t, err)
+
+		pkg, diags, err := schema.BindSpec(spec, nil)
+		assert.NoError(t, err)
+		assert.Empty(t, diags)
+		assert.NotNil(t, pkg)
+
+		//Kludge
```
Comment (Frassle) [inline pkg/testing/matrix/matrix.go:226]: Not needed anymore?

File: pkg/testing/matrix/matrix.go [hunk 10]
```diff
@@ -0,0 +1,457 @@
...
+	}
+
+	//Execute build commands and add plugin links
+	for _, plugin := range opts.Plugins {
+		for _, cmd := range plugin.Build {
+			err := cmd.Run()
+			assert.NoError(t, err)
+		}
+
+		localProjects[plugin.Name] = fmt.Sprintf("%s/%s-sdk", dir, plugin.Name)
```
Comment (Frassle) [inline pkg/testing/matrix/matrix.go:152]: I think we want to do this per langauge, we know here that we're going to build SDKs into "/dotnet" folders etc because we do that explicitly below. No need to assume that all dotnet SDKs are inside a "dotnet" folder inside GenerateProgram.

File: pkg/testing/matrix/matrix.go [hunk 11]
```diff
@@ -0,0 +1,457 @@
...
+				err = NodeConfigurePkg(pkg)
+				assert.NoError(t, err)
+				files, err = jsgen.GeneratePackage(pkgName, pkg, files)
+				assert.NoError(t, err)
+			case DOTNET:
+				err = DotnetConfigurePkg(pkg)
+				assert.NoError(t, err)
+				files, err = dotnetgen.GeneratePackage(pkgName, pkg, files)
+				assert.NoError(t, err)
+			//In the future we should support java but I hava no idea where to even start with that.
```
Comment (Frassle) [inline pkg/testing/matrix/matrix.go:261]: That's fine we can circle back to java later

File: pkg/testing/matrix/matrix.go [hunk 12]
```diff
@@ -0,0 +1,457 @@
...
+			default:
+				fmt.Printf("Unknown language: '%s'", lang)
+				continue
+			}
+
+			sdkDir := filepath.Join(dir, fmt.Sprintf("%s-sdk", pkgName), lang)
+			for p, file := range files {
+				// TODO: full conversion from path to filepath
```
Comment (Frassle) [inline pkg/testing/matrix/matrix.go:269]: Still TODO?

File: pkg/testing/matrix/matrix.go [hunk 13]
```diff
@@ -0,0 +1,457 @@
...
+				err = os.MkdirAll(filepath.Join(sdkDir, path.Dir(p)), 0700)
+				assert.NoError(t, err)
+				err = os.WriteFile(filepath.Join(sdkDir, p), file, 0600)
+				assert.NoError(t, err)
+			}
+			if lang == "go" {
```
Comment (Frassle) [inline pkg/testing/matrix/matrix.go:275]: Why are these needed? This feels like something GeneratePackage ought to do? Do we need a new "generateOnly" flag to GeneratePackage so that this logic can live in the language code not here?

File: pkg/testing/matrix/matrix.go [hunk 14]
```diff
@@ -0,0 +1,457 @@
...
+		})*/
+	}
+}
+
+func PythonConfigurePkg(pkg *schema.Package) error {
+	pkg.Version = &semver.Version{
+		Major: pkg.Version.Major,
+		Minor: pkg.Version.Minor,
+		Patch: pkg.Version.Patch,
+	} // Prune patch and build versions, since semver and python don't seem to agree on formatting.
```
Comment (Frassle) [inline pkg/testing/matrix/matrix.go:389]: Hah they don't, playing with pulumictl for this over at https://github.com/pulumi/pulumictl/pull/51

File: pkg/testing/matrix/matrix.go [hunk 15]
```diff
@@ -0,0 +1,457 @@
...
+	} else if notnil && !ismessage {
+		info, ok := raw.(jsgen.NodePackageInfo)
+		if !ok {
+			return fmt.Errorf("invalid nodejs package info")
+		}
+		nodePkg = info
+	} else if !notnil {
+		nodePkg = jsgen.NodePackageInfo{}
+	}
+	nodePkg.RespectSchemaVersion = true
```
Comment (Frassle) [inline pkg/testing/matrix/matrix.go:431]: Why do node and python need this but nothing else? That seems inconsistent.

File: pkg/testing/matrix/matrix.go [hunk 16]
```diff
@@ -0,0 +1,457 @@
...
+		langPtOpts = langPtOpts.With(i.ProgramTestOptions{
+			Dir: subdir,
+		})
+		t.Run(langOpt.Language, func(t *testing.T) {
+			i.ProgramTest(t, &langPtOpts)
+		})
+		//clean up subdir
+
+		//Including this seems to cause all tests to fail.
+		/*t.Cleanup(func() {
```
Comment (Frassle) [inline pkg/testing/matrix/matrix.go:378]: Odd? What was the error?

File: pkg/testing/matrix/matrix_test.go [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+
+	"github.com/blang/semver"
+	i "github.com/pulumi/pulumi/pkg/v3/testing/integration"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+)
+
+func TestAll(t *testing.T) {
+	t.Parallel()
+	opts := []TestOptions{
+		/*{
```
Comment (Frassle) [inline pkg/testing/matrix/matrix_test.go:29]: Can these all be enabled now?

File: pkg/testing/matrix/matrix_test.go [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+						Initialized: false,
+					},
+				},
+			},
+			Languages: allLanguages(),
+		},*/
+		{
+			Program: &i.ProgramTestOptions{
+				Dir:              "tests/provider",
+				PulumiBin:        "/Users/hliuson/.pulumi-dev/bin/pulumi",
```
Comment (Frassle) [inline pkg/testing/matrix/matrix_test.go:62]: Probably leave this unset for now, we might want to look at having the makefiles go build to ./bin so we can hardcode to that path here maybe 
(golangs lets just dump everything across all projects in to one big ~/go folder by default is a mad design)

File: pkg/testing/matrix/tests/empty/Pulumi.yaml [hunk 19]
```diff
@@ -0,0 +1,4 @@
+name: empty
```
Comment (Frassle) [inline pkg/testing/matrix/tests/empty/Pulumi.yaml:1]: The .dev.yaml files shouldn't need to be checked in.

File: sdk/go/common/workspace/plugins.go [hunk 20]
```diff
@@ -1342,11 +1342,9 @@ func getPluginInfoAndPath(
 		for j, p2 := range projectPlugins {
 			if j < i {
 				if p2.Kind == p1.Kind && p2.Name == p1.Name {
-					if p1.Version != nil && p2.Version != nil && p2.Version.Equals(*p1.Version) {
-						return nil, "", fmt.Errorf(
-							"multiple project plugins with kind %s, name %s, version %s",
-							p1.Kind, p1.Name, p1.Version)
-					}
+					return nil, "", fmt.Errorf(
```
Comment (Frassle) [inline sdk/go/common/workspace/plugins.go:1345]: Not allowing aws@v4 and aws@v5 anymore? Per version stuff seemed to be working ok before.

File: pkg/codegen/go/gen_program.go [hunk 21]
```diff
@@ -145,21 +146,34 @@ func GenerateProject(directory string, project workspace.Project, program *pcl.P
...
+	if err != nil {
+		return err
+	}
+
 	// Build a go.mod based on the packages used by program
 	var gomod bytes.Buffer
 	gomod.WriteString("module " + project.Name.String() + "\n")
+	gomod.WriteString("go 1.17\n")
+	for name, sdk := range localProjects {
+		if name == "go" {
```
[the remark this answers is not in the document]
Comment (harribo11) [reply pkg/codegen/go/gen_program.go:159]: I've opted to use the language instead of "pulumi" for the language SDKs, because otherwise either 1. "pulumi" would point to different places for different tests (doable, imo slightly confusing but not the worst), or 2. "pulumi" would just point to p/p/sdk and then the rest of the path would have to be appended (makes assumptions about where the SDKs are). I think that it makes more sense to just do it this way, particularly if developing only one language SDK.

[the remark this answers is not in the document]
Comment (Frassle) [reply pkg/codegen/go/gen_program.go:159]: Ah right but see comments further down that I think the matrix testing should be building this map per-language, and so both of these issues would be solved

File: pkg/testing/matrix/matrix.go (hunk 7 above)
Comment (harribo11) [reply pkg/testing/matrix/matrix.go:137]: The issue is that we need to have plugin links in the base YAML sometimes otherwise it can't be read into the program.

File: pkg/testing/matrix/matrix_test.go (hunk 17 above)
Comment (harribo11) [reply pkg/testing/matrix/matrix_test.go:29]: Stack references still cause the whole thing to panic, otherwise yes

File: sdk/go/common/workspace/plugins.go (hunk 20 above)
Comment (harribo11) [reply sdk/go/common/workspace/plugins.go:1345]: Maybe just match major versions?. When I was testing, the exact version matching was fairly brittle and would often cause the plugin to be missed.

Comment (harribo11) [reply sdk/go/common/workspace/plugins.go:1345]: That can be a separate PR though

File: pkg/testing/matrix/matrix.go (hunk 16 above)
Comment (harribo11) [reply pkg/testing/matrix/matrix.go:371]: Seems to have resolved itself

File: pkg/testing/matrix/matrix.go (hunk 13 above)
Comment (harribo11) [reply pkg/testing/matrix/matrix.go:268]: I added this at some point but I don't remember why! Seems to work okay if I remove it.

File: pkg/testing/matrix/matrix.go (hunk 15 above)
Comment (harribo11) [reply pkg/testing/matrix/matrix.go:424]: Otherwise it just says ${VERSION} and pip/node freak out

File: pkg/codegen/dotnet/gen_program.go [hunk 22]
```diff
@@ -157,7 +158,17 @@ func GenerateProgram(program *pcl.Program) (map[string][]byte, hcl.Diagnostics,
...
+	local, ok := localProjects[pkg]
+	if !ok {
+		return "", false
+	}
+	dotnet, ok := local["dotnet"]
+	return dotnet, ok
+}
+
+func GenerateProject(directory string, project workspace.Project, program *pcl.Program,
+	localProjects map[string]map[string]string) error {
```
Comment (Frassle) [inline pkg/codegen/dotnet/gen_program.go:171]:
```suggestion
	localProjects map[string]string) error {
```
Resolve to the "dotnet" map in the caller of this method. There's no point passing a map here that we're only going to look at the "dotnet" key in.

File: pkg/testing/matrix/matrix.go (hunk 7 above)
Comment (Frassle) [reply pkg/testing/matrix/matrix.go:137]: Right, should we not maintain those links though?

Comment (harribo11) [reply pkg/testing/matrix/matrix.go]: Ah, figured out how to use the specified plugins in code to load from YAML. So now plugins should only be specified in one place.

File: sdk/go/common/workspace/plugins.go (hunk 20 above)
Comment (AaronFriel) [reply sdk/go/common/workspace/plugins.go:1395]: This looks like it would affect Pulumi YAML and plugin resolution in general.

Comment (harribo11) [reply sdk/go/common/workspace/plugins.go:1395]: This specific check only looks at linked plugins - it was introduced in #10146

File: pkg/codegen/dotnet/gen_program.go [hunk 23]
```diff
@@ -185,34 +187,55 @@ func GenerateProject(directory string, project workspace.Project, program *pcl.P
...
 
 	<ItemGroup>
-		<PackageReference Include="Pulumi" Version="3.*" />
 `)
 
-	// For each package add a PackageReference line
+	packageTemplate := "		<PackageReference Include=\"%s\" Version=\"%s\" />\n"
+	projectTemplate := "		<ProjectReference Include=\"%s\" />\n"
+
+	if localProjects == nil {
```
Comment (Frassle) [inline pkg/codegen/dotnet/gen_program.go:195]: This isn't needed? A lookup in a nil map just returns nil

File: pkg/codegen/go/gen_program.go [hunk 24]
```diff
@@ -153,21 +154,35 @@ func GenerateProject(directory string, project workspace.Project, program *pcl.P
...
```
Comment (Frassle) [inline pkg/codegen/go/gen_program.go:167]:
```suggestion
		if name == "pulumi" {
```

File: pkg/codegen/nodejs/gen_program.go [hunk 25]
```diff
@@ -155,16 +157,24 @@ func GenerateProject(directory string, project workspace.Project, program *pcl.P
 	}
 	files["Pulumi.yaml"] = projectBytes
 
+	if localProjects == nil {
+		localProjects = map[string]string{}
+	}
+
 	// Build the pacakge.json
 	var packageJSON bytes.Buffer
+	pulumi := `@pulumi/pulumi": "^3.0.0`
```
Comment (Frassle) [inline pkg/codegen/nodejs/gen_program.go:166]:
```suggestion
	pulumi := `"@pulumi/pulumi": "^3.0.0"`
```
I thought this was a bugged at first glance, simpler to keep all the quotes together rather than wrapping the outer quotes as part of the fmt string below.

File: pkg/codegen/nodejs/gen_program.go [hunk 26]
```diff
@@ -155,16 +157,24 @@ func GenerateProject(directory string, project workspace.Project, program *pcl.P
...
+	if localPulumi, ok := localProjects["pulumi"]; ok {
+		pulumi = fmt.Sprintf("@pulumi/pulumi\": \"file:%s", localPulumi)
```
Comment (Frassle) [inline pkg/codegen/nodejs/gen_program.go:168]:
```suggestion
		pulumi = fmt.Sprintf("\"@pulumi/pulumi\": \"file:%s\"", localPulumi)
```

File: pkg/codegen/nodejs/gen_program.go [hunk 27]
```diff
@@ -155,16 +157,24 @@ func GenerateProject(directory string, project workspace.Project, program *pcl.P
...
+	}
 	packageJSON.WriteString(fmt.Sprintf(`{
 		"name": "%s",
 		"devDependencies": {
 			"@types/node": "^14"
 		},
 		"dependencies": {
 			"typescript": "^4.0.0",
-			"@pulumi/pulumi": "^3.0.0"`, project.Name.String()))
+			"%s"`, project.Name.String(), pulumi))
```
Comment (Frassle) [inline pkg/codegen/nodejs/gen_program.go:177]:
```suggestion
			%s`, project.Name.String(), pulumi))
```

File: pkg/testing/matrix/matrix.go [hunk 28]
```diff
@@ -226,7 +226,12 @@ func Test(t *testing.T, opts TestOptions) {
 		assert.NoError(t, err)
 		assert.NotNil(t, provider)
 
-		schemaBytes, err := provider.GetSchema(int(plugin.Version.Major))
+		max := plugin.Version.Major
+		if max > 1 {
+			max = 1
+		}
+
+		schemaBytes, err := provider.GetSchema(int(max))
```
Comment (Frassle) [inline pkg/testing/matrix/matrix.go:234]: Isn't this just always 1?

Comment (harribo11) [reply pkg/testing/matrix/matrix.go:234]: I think it can be zero? Otherwise why would GetSchema take any params at all?

Comment (Frassle) [reply pkg/testing/matrix/matrix.go:234]: @AaronFriel or @iwahbe might know this for sure

**Comment by harribo11** [comment]
See https://github.com/pulumi/pulumi-digitalocean/pull/328/ for an example use of this.

File: pkg/testing/matrix/matrix.go (hunk 28 above)
Comment (iwahbe) [reply pkg/testing/matrix/matrix.go:170]: As far as I'm aware, it is always 0. This is how we generally call `GetSchema`:
https://github.com/pulumi/pulumi/blob/f618cf8c61a4df51e5e9a2741b3b5d05bbdd21d8/pkg/codegen/schema/loader.go#L246-L247

I assume that the parameter is just to allow us to future proof against breaking schema changes.

**Comment by harribo11** [comment]
Outstanding Issues: @stack72 

- CI currently failing tests when fetching plugins, not entirely sure why. This might be one that we want to cut as it introduces external dependencies into tests. 
- May want to give users some configuration options, such as where to put the generated files, whether or not to cleanup afterwards, etc.
- Current method of building plugins is somewhat janky and we may just want to leave that to the user.
- Some languages fail pretty simple codegen tests, (see tests/scalar and tests/structured). tests/reference was previously broken but I think #10419 fixed it.
- A couple random //nolint statements

**Review by danielrbradley — COMMENTED** [review]
I'd recommend closing this as this specific implementation of matrix testing is now unlikely to land due to staleness.
